File: hasher.py
"""
哈希计算模块
-----------
两级哈希策略：
  quick_hash — 只读文件头 8KB，速度极快，用于快速排除不同内容文件
  full_hash  — 流式读取完整文件，精确但慢，仅对 quick_hash 碰撞的候选文件执行
使用 XXH128 算法，速度极快(>10GB/s)，128位碰撞概率极低。
"""

import os
import xxhash
from config import QUICK_HASH_HEAD_BYTES, QUICK_HASH_TAIL_BYTES, HASH_READ_CHUNK


def quick_hash(filepath):
    """Compute XXH128 hash of head + tail bytes of a file.

    Returns bytes (16 bytes for XXH128), or None if file can't be read.
    """
    if not os.path.exists(filepath):
        return None
    try:
        file_size = os.path.getsize(filepath)
    except OSError:
        return None

    h = xxhash.xxh128()
    head_size = min(QUICK_HASH_HEAD_BYTES, file_size)
    tail_size = min(QUICK_HASH_TAIL_BYTES, file_size - head_size)

    try:
        with open(filepath, "rb") as f:
            # Read head
            if head_size > 0:
                _hash_stream(f, head_size, h)

            # Seek to tail
            if tail_size > 0:
                f.seek(-tail_size, os.SEEK_END)
                _hash_stream(f, tail_size, h)

        return h.digest()
    except (OSError, PermissionError, FileNotFoundError):
        return None


def full_hash(filepath):
    """Compute XXH128 hash of the entire file (streaming).

    Returns bytes (16 bytes), or None if file can't be read.
    """
    if not os.path.exists(filepath):
        return None
    h = xxhash.xxh128()
    try:
        with open(filepath, "rb") as f:
            _hash_stream(f, float("inf"), h)
        return h.digest()
    except (OSError, PermissionError, FileNotFoundError):
        return None


def _hash_stream(f, max_bytes, hasher):
    """Read up to max_bytes from file f into hasher."""
    remaining = max_bytes
    while remaining > 0:
        chunk_size = min(HASH_READ_CHUNK, remaining)
        data = f.read(chunk_size)
        if not data:
            break
        hasher.update(data)
        remaining -= len(data)

File: c-sharp-override/Models/Hasher.cs
using K4os.Hash.xxHash;

namespace CleanDup.Models;

/// <summary>
/// Hash computation using XXH128 (K4os.Hash.xxHash).
/// quick_hash: reads head + tail bytes only (fast filter).
/// full_hash: streaming read of the entire file.
/// </summary>
public static class Hasher
{
    /// <summary>XXH128 hash of head + tail bytes. Returns 16 bytes or null.</summary>
    public static byte[]? QuickHash(string filePath)
    {
        if (!File.Exists(filePath)) return null;
        try
        {
            var fileSize = new FileInfo(filePath).Length;
            var hasher = XXH128.DigestOf(Array.Empty<byte>()); // placehold, we'll build incrementally

            using var fs = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
            var headSize = (int)Math.Min(Config.QuickHashHeadBytes, fileSize);
            var tailSize = (int)Math.Min(Config.QuickHashTailBytes, fileSize - headSize);

            // Read head
            if (headSize > 0)
            {
                var buf = new byte[headSize];
                fs.ReadExactly(buf);
                hasher = XXH128.DigestOf(buf);
            }

            // Seek to tail and re-hash combined
            if (tailSize > 0)
            {
                fs.Seek(-tailSize, SeekOrigin.End);
                var tailBuf = new byte[tailSize];
                fs.ReadExactly(tailBuf);

                // Combine head + tail into one hash
                var combined = new byte[headSize + tailSize];
                if (headSize > 0)
                {
                    fs.Seek(0, SeekOrigin.Begin);
                    fs.ReadExactly(combined, 0, headSize);
                }
                Array.Copy(tailBuf, 0, combined, headSize, tailSize);
                return XXH128.DigestOf(combined);
            }

            return hasher;
        }
        catch { return null; }
    }

    /// <summary>XXH128 hash of the entire file (streaming).</summary>
    public static byte[]? FullHash(string filePath)
    {
        if (!File.Exists(filePath)) return null;
        try
        {
            using var fs = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
            var buf = new byte[Config.HashReadChunk];
            var hasher = new XXH128();
            int read;
            while ((read = fs.Read(buf, 0, buf.Length)) > 0)
                hasher.Update(buf.AsSpan(0, read));
            return hasher.Digest();
        }
        catch { return null; }
    }
}
